File: 프로그래머스/1/172928. 공원 산책/공원 산책.py

def solution(park, routes):
    
    coord = [0,0]
    max_x = len(park[0]) - 1
    max_y = len(park) - 1
    
    direction = {
        'N': [-1, 0],
        'S': [1, 0],
        'W': [0, -1],
        'E': [0,1]
    }
    
    for i in range(len(park)):
        j = park[i].find('S')
        if(j == -1):
            continue;
        coord = [i, j]
    for i in routes:
        d = i.split(" ")[0]
        c = int(i.split(" ")[1])
        
        if (coord[0] + (direction[d][0] * (c))) > max_y or (coord[0] + (direction[d][0] * (c))) < 0:
            continue
        if (coord[1] + (direction[d][1] * (c))) > max_x or (coord[1] + (direction[d][1] * (c))) < 0:
            continue
        k = 0
        for j in range(1, c+1):
            if park[coord[0] + (direction[d][0] * j)][coord[1] + direction[d][1] * j] == 'X':
                break
            k = k+1
        if(k == c):
            coord = [coord[0] + (direction[d][0] * j), coord[1] + (direction[d][1] * j)]
        
            
                     
    return coord

File: 프로그래머스/1/172928. 공원 산책/공원 산책.js
function isOut(w, h, x ,y){
    return y >= w || y < 0 || x >= h || x < 0
}

function hasObstacle(park, sx, sy, dx, dy, step){
    if(step+1 == 0) return 0
    if(park[sx][sy] == 'X') return 1
    return hasObstacle(park, sx + dx, sy + dy, dx, dy, step - 1 )
}

function solution(park, routes) {
    var answer = [];
    
    const width = park[0].length
    const height = park.length
    
    let [sx, sy] = [Math.floor(park.join("").indexOf("S")/width), park.join("").indexOf("S")%width]
    for(let i in routes){
        const direction = routes[i].split(" ")[0]
        const step = Number(routes[i].split(" ")[1])
        
        if(direction == 'N'){
            if(isOut(width, height, sx - step, sy) || hasObstacle(park, sx, sy, -1, 0, step)) continue
            sx -= step
        }
        if(direction == 'S'){
            if(isOut(width, height, sx + step, sy) || hasObstacle(park, sx, sy, 1, 0, step)) continue
            sx += step
        }
        if(direction == 'W'){
            if(isOut(width, height, sx, sy - step) || hasObstacle(park, sx, sy, 0, -1, step)) continue
            sy -= step
        }
        if(direction == 'E'){
            if(isOut(width, height, sx, sy + step) || hasObstacle(park, sx, sy, 0, 1, step)) continue
            sy += step
        }
        
    }
    
    answer = [sx, sy]
    
    return answer;
}